### PR TITLE
Use tracing crate to have more context in log messages.

### PR DESCRIPTION
**Pull request by hlinnaka**
Whenever we start processing a request, we now enter a tracing "span"
that includes context information like the tenant and timeline ID,
and the operation we're performing. That context information gets
attached to every log message we create within the span. That way, we
don't need to include basic context information like that in every log
message, and it also becomes easier to filter the logs programmatically.

This removes the eplicit timeline and tenant IDs from most log messages,
as you get that information from the enclosing span now.

Also improve log messages in general, dialing down the level of some
messages that are not very useful, and adding information to others.

We now obey the RUST_LOG env variable, if it's set.

The 'tracing' crate allows for different log formatters, like JSON or
bunyan output. The one we use now is human-readable, with ANSI colors,
pretty close to the old format.

### CODE REVIEW

**Comment by hlinnaka** [comment]
@SomeoneToIgnore, you mentioned in the all hands today that you had started hacking on this. As it happens, so did I a couple of weeks ago but never got around to finish it into a PR. Here's what I ended up with, does this look similar to what you had in mind?

**Comment by hlinnaka** [comment]
We have some options on the output log format. This currently prints something very similar to what we have in 'main today:
```
Oct 01 21:31:58.052 TRACE pagestream{timeline=4b1f73f3b2fe5c20727ab3532deaf3ee tenant=b044e526cebb15ee983d648614127f1a}: query: CopyData(b"\x02\x01\0\0\0\0\x01k\x1c\xc0\0\0\x06\x80\0\0\0\0\0\0\x04\xd1\0\0\0\0\x01")
Oct 01 21:31:58.052 TRACE pagestream{timeline=4b1f73f3b2fe5c20727ab3532deaf3ee tenant=b044e526cebb15ee983d648614127f1a}:get_page{rel=1664/0/1233 blkno=1 req_lsn=0/16B1CC0}: get_layer_for_read called for 1664/0/1233.0 at 0/16B1CC0
Oct 01 21:31:58.052 TRACE pagestream{timeline=4b1f73f3b2fe5c20727ab3532deaf3ee tenant=b044e526cebb15ee983d648614127f1a}:get_page{rel=1664/0/1233 blkno=1 req_lsn=0/16B1CC0}: found layer in cache: 4b1f73f3b2fe5c20727ab3532deaf3ee 0/169C350-0/169C351
Oct 01 21:31:58.052 TRACE pagestream{timeline=4b1f73f3b2fe5c20727ab3532deaf3ee tenant=b044e526cebb15ee983d648614127f1a}:get_page{rel=1664/0/1233 blkno=1 req_lsn=0/16B1CC0}: found page image for blk 1 in 1664/0/1233 at 0/16B1CC0, no WAL redo required
Oct 01 21:31:58.052 TRACE pagestream{timeline=4b1f73f3b2fe5c20727ab3532deaf3ee tenant=b044e526cebb15ee983d648614127f1a}: query: CopyData(b"\x02\x01\0\0\0\0\x01k\x1c\xc0\0\0\x06\x80\0\0\0\0\0\0\x04\xbe\0\0\0\0\0")
Oct 01 21:31:58.053 TRACE pagestream{timeline=4b1f73f3b2fe5c20727ab3532deaf3ee tenant=b044e526cebb15ee983d648614127f1a}:get_page{rel=1664/0/1214 blkno=0 req_lsn=0/16B1CC0}: get_layer_for_read called for 1664/0/1214.0 at 0/16B1CC0
Oct 01 21:31:58.053 TRACE pagestream{timeline=4b1f73f3b2fe5c20727ab3532deaf3ee tenant=b044e526cebb15ee983d648614127f1a}:get_page{rel=1664/0/1214 blkno=0 req_lsn=0/16B1CC0}: found layer in cache: 4b1f73f3b2fe5c20727ab3532deaf3ee 0/169C350-0/169C351
Oct 01 21:31:58.053 DEBUG pagestream{timeline=4b1f73f3b2fe5c20727ab3532deaf3ee tenant=b044e526cebb15ee983d648614127f1a}:get_page{rel=1664/0/1214 blkno=0 req_lsn=0/16B1CC0}: loaded from ./tenants/b044e526cebb15ee983d648614127f1a/timelines/4b1f73f3b2fe5c20727ab3532deaf3ee/rel_1664_0_1214_0_0_000000000169C350    
Oct 01 21:31:58.053 TRACE pagestream{timeline=4b1f73f3b2fe5c20727ab3532deaf3ee tenant=b044e526cebb15ee983d648614127f1a}:get_page{rel=1664/0/1214 blkno=0 req_lsn=0/16B1CC0}: found page image for blk 0 in 1664/0/1214 at 0/16B1CC0, no WAL redo required
```

The lines are very wide, which is not very nice on the terminal. But it's important to have the context there, and when you print the tenant and timeline id it just takes a lot of space, there's no way around that.

Alternatively, we could switch to a more structured JSON format, the 'tracing' crate comes with a built-in formatter for that. I'm not sure what the state of the art with that is, but presumably there are some nice tools out there to parse and display JSON logs. Anyone have experience with that?

File: pageserver/src/page_service.rs [hunk 1]
```diff
@@ -363,6 +365,8 @@ impl PageServerHandler {
         timeline: &dyn Timeline,
         req: &PagestreamExistsRequest,
     ) -> Result<PagestreamBeMessage> {
+        let _enter = info_span!("get_rel_exists", rel = %req.rel, req_lsn = %req.lsn).entered();
```
Comment (SomeoneToIgnore) [inline pageserver/src/page_service.rs:368]: I'm not a big fan of proc macros myself, so not recommending to do so, but just to show an alternative to such span creation: https://github.com/tokio-rs/tracing/blob/e0db0563064d4bb0602ac7981e4ed8fa7692b603/tracing-attributes/tests/fields.rs#L9

When an annotation like this placed, the method/function body gets wrapped with the tracing span with name = fn name, all parameters logged (unless excluded explicitly), some magical syntax to derive the vars.

File: zenith_utils/Cargo.toml [hunk 2]
```diff
@@ -18,12 +18,9 @@ serde = { version = "1.0", features = ["derive"] }
...
 thiserror = "1.0"
 tokio = "1.11"
-
-slog-async = "2.6.0"
-slog-stdlog = "4.1.0"
-slog-scope = "4.4.0"
-slog-term = "2.8.0"
-slog = "2.7.0"
+tracing = "0.1.27"
+tracing-log = "0.1.2"
```
Comment (SomeoneToIgnore) [inline zenith_utils/Cargo.toml:22]: See the first bullet in 
https://docs.rs/tracing-subscriber/0.2.24/tracing_subscriber/#optional-dependencies

tracing-subscriber with default features is enough to init tracing backend with `log` redirected inside it, see my branch for details.

**Review by SomeoneToIgnore — COMMENTED** [review]
Here's my branch: https://github.com/zenithdb/zenith/compare/kb/attempt-to-tracing?expand=1
It's pretty small and concentrated on the "right" way to do things in `logging.rs` lib file, so pick whatever you find useful and let me know when I can remove the branch.

tracing is somewhat odd when in comes to configuring, because there's so many moving parts that could be put together many different ways.
In my case, I've used a more formalized way to allow writing without blocks (supposed to be good, but needs benches and comparison) + rotate the files automatically (cool, not sure if super needed).

To sum this up:
* I like how simple `with_writer` setup is in ypur PR, I wonder, if we could try the same with the non-blocking option and check if it works and works better
* we need to treat async code a bit more gently or do not place spans (logging is fine) inside that code
* a few extra deps in Cargo.toml are worth throwing off

File: zenith_utils/src/logging.rs [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
 };
 
 use anyhow::{Context, Result};
 
-pub fn init(
-    log_filename: impl AsRef<Path>,
-    daemonize: bool,
-) -> Result<(slog_scope::GlobalLoggerGuard, File)> {
+use tracing::subscriber::set_global_default;
+//use tracing_bunyan_formatter::{BunyanFormattingLayer, JsonStorageLayer};
```
Comment (sharnoff) [inline zenith_utils/src/logging.rs:9]: Remove this comment?

**Review by sharnoff — APPROVED** [review]
Dang, nice! Nothing really to say that @SomeoneToIgnore hasn't already mentioned, just weighing in with my two cents on holding `.entered()` guards at the top of functions.

IMO it looks a little weird. Proc macros have their downsides, but I can't think of any other clean ways to do it -- it's maybe worth putting some thought into it.

File: zenith_utils/src/logging.rs [hunk 4]
```diff
@@ -18,30 +20,43 @@ pub fn init(
...
+
+        let fmt_layer = fmt::layer()
+            .pretty()
+            .with_target(false) // don't include event targets when logging
+            // Don't use colors in log file.
+            //
+            // FIXME: It's actually nice to have them, so leave them on for now. Use
+            // 'less -R' option to display them.
+            //
+            // .with_ansi(false)
```
[the remark this answers is not in the document]
Comment (hlinnaka) [reply zenith_utils/src/logging.rs:46]: Makes sense, changed it to non-ansi.

However, that ran into a bug in 'tracing'. I'm currently actually using the 'pretty' format, so the log looks like this:
```
  Oct 07 13:18:22.704  INFO GET /v1/branch/40d00a53756334d1a471043acce25557/test_config 200 OK
    at zenith_utils/src/http/endpoint.rs:26

  Oct 07 13:18:22.732  INFO taking basebackup lsn=0/0, prev_lsn=0/169C350
    at pageserver/src/basebackup.rs:84
    in basebackup with ESC[1mtimelineESC[0m: 1d97ed9047930b55c1c4535db09ec4f5, ESC[1mtenantESC[0m: 40d00a53756334d1a471043acce25557

  Oct 07 13:18:22.733  INFO list_nonrels called at 0/169C350
    at pageserver/src/layered_repository.rs:725
    in basebackup with ESC[1mtimelineESC[0m: 1d97ed9047930b55c1c4535db09ec4f5, ESC[1mtenantESC[0m: 40d00a53756334d1a471043acce25557, ESC[1mlsnESC[0m: "0/169C350"
```
So even with `with_ansi(false)`, there are still some ansi control chars in the context lines. There's a github issue about that at https://github.com/tokio-rs/tracing/issues/1310. I propose that we go ahead and push this even though we still get some of those control chars, and wait/work on the upstream issue to be fixed.

File: pageserver/src/http/routes.rs [hunk 5]
```diff
@@ -144,10 +148,13 @@ async fn tenant_list_handler(request: Request<Body>) -> Result<Response<Body>, A
     // check for management permission
     check_permission(&request, None)?;
 
-    let response_data =
-        tokio::task::spawn_blocking(move || crate::branches::get_tenants(get_config(&request)))
-            .await
-            .map_err(ApiError::from_err)??;
+    let response_data = tokio::task::spawn_blocking(move || {
+        let _enter = info_span!("tenant_list").entered();
```
[the remark this answers is not in the document]
Comment (hlinnaka) [reply pageserver/src/http/routes.rs:152]: Can you elaborate, do you see an active bug in how this PR uses spans, or are you just pointing out that we have to be careful with async?

File: zenith_utils/src/logging.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply zenith_utils/src/logging.rs]: I like the idea to proceed anyway.
What if we disable the pretty logging too, at least for now? 

----
What does that `pretty` mode do, by the way? I've not used it before 😄
Docs say that it should be "pretty": https://docs.rs/tracing-subscriber/0.2.25/tracing_subscriber/fmt/struct.Layer.html#method.pretty

> Sets the layer being built to use an excessively pretty, human-readable formatter.

which is not much, grepping along the repo shows this:
https://github.com/tokio-rs/tracing/blob/6b7051c9897a6ea2afd61cfa139714cfeed8bf4f/tracing-subscriber/src/fmt/mod.rs#L120

```rust
//! * [`format::Pretty`]: Emits excessively pretty, multi-line logs, optimized
//!   for human readability. This is primarily intended to be used in local
//!   development and debugging, or for command-line applications, where
//!   automated analysis and compact storage of logs is less of a priority than
//!   readability and visual appeal.
```

If that's really the case, I don't like it:
* it's not pretty right now
* multiline logging is often tricky to handle in production cloud environment, when tons of threads concurrently writing something: unless every that line is prepended with some id/guid/uuid/whatnot , it's impossible to distinguish one line from another. We can dare and explore tracing behavior, or we can prefer single lines everywhere.

File: zenith_utils/Cargo.toml (hunk 2 above)
Comment (SomeoneToIgnore) [reply zenith_utils/Cargo.toml]: Good catch, I've missed that part too, because had called `.finish();` on my subscriber builder...
Like I've mentioned, the configuration is somewhat convoluted here.

Here's the difference in the in their sources: https://github.com/tokio-rs/tracing/blob/6b7051c9897a6ea2afd61cfa139714cfeed8bf4f/tracing-subscriber/src/fmt/mod.rs#L556-L579

the main bit is

```rust
/// If the `tracing-log` feature is enabled, this will also install
/// the LogTracer to convert `Log` records into `tracing` `Event`s.
```

that feature is enabled with the default features, so you should have it, then you should either invoke the right method on the formatter, or do this manually:
https://github.com/tokio-rs/tracing/blob/master/tracing-subscriber/src/util.rs#L57-L67

File: pageserver/Cargo.toml [hunk 6]
```diff
@@ -35,6 +35,7 @@ scopeguard = "1.1.0"
 rust-s3 = { version = "0.27.0-rc4", features = ["no-verify-ssl"] }
 async-trait = "0.1"
 const_format = "0.2.21"
+tracing = "0.1.27"
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/Cargo.toml:38]: That did not get removed though? 

All `log::debug!` calls should be done as `tracing::debug!` with the new approach, and tracing has its own counterparts for log levels and other structs to replace the log ones, so we should not rely on the crate in our project still?

That "missing log" case discussed separately also benefits from using `tracing` instead of `log` instead 😄
